### ISSUE TITLE
we find a issue, I don't konw is there a bug?

### ISSUE THREAD

**Issue by driftluo** [rust-lang/rust#49865]
the function like this is ok:
```rust
fn test() -> bool {
    let a = {true} && true;
    a 
}
```

but when is become this, there is an error:
```rust
fn test() -> bool {
    {
    // this is some thing return bool
    true
    } && true
}
```
the error is:
```
error[E0308]: mismatched types
 --> src/main.rs:4:5
  |
4 |     true
  |     ^^^^ expected (), found bool
  |
  = note: expected type `()`
             found type `bool`

error[E0308]: mismatched types
 --> src/main.rs:6:7
  |
1 | fn test() -> bool {
  |              ---- expected `bool` because of return type
...
6 |     } && true
  |       ^^^^^^^ expected bool, found &&bool
  |
  = note: expected type `bool`
             found type `&&bool`

error: aborting due to 2 previous errors
```

**Comment by petrochenkov** [comment]
Yep, this is expected behavior.
Disambiguation is done in favor of two statements to support things like this
```rust
if cond {
    stmts;
}

*ptr = val; // Not multiplication!
```
Yes, there are false positives, but the opposite disambiguation has even more false positives.

**Comment by petrochenkov** [comment]
@driftluo 
Because it's interpreted as two statements
```rust
// Statement 1
{
    true
}

// Statement 2
&&true // Take address of address of `true`, the result has type `&&bool`
```

**Comment by driftluo** [comment]
This means that if there is a braces block, it will be interpreted as a single statement regardless of what follows.
If I want it to a expression, I should use () to tell rustc, it is expression.

**Comment by crlf0710** [comment]
@petrochenkov while the decision is understandable, the diagnostics is far from satisfying... Any chance to improve it somehow?

**Comment by ZhangHanDong** [comment]
@crlf0710  Agree

**Comment by KiChjang** [comment]
This issue has been discussed in #29071.
